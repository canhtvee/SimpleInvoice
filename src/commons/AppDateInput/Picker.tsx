import React from 'react';
import DatePicker, {DatePickerProps} from 'react-native-date-picker';

export interface PickerProps extends Omit<DatePickerProps, 'open'> {
  ref: (instance: any) => void;
}

export const Picker = React.forwardRef((props: PickerProps, ref: any) => {
  const [openModal, setOpenModal] = React.useState(false);

  React.useImperativeHandle(ref, () => ({
    show: () => setOpenModal(true),
    hide: () => setOpenModal(false),
  }));

  return <DatePicker open={openModal} {...props} />;
});
