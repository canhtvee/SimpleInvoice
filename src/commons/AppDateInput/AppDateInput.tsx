import React, {useRef, useState} from 'react';
import {
  StyleSheet,
  Text,
  View,
  StyleProp,
  ViewStyle,
  TextStyle,
  Pressable,
} from 'react-native';
import {Controller, useFormState, UseControllerProps} from 'react-hook-form';
import dayjs from 'dayjs';

import {Colors, Sizes} from '../../utils';

import {AppText} from '../appText';

import {Picker, PickerProps} from './Picker';
import DatePicker from 'react-native-date-picker';

export interface AppDateInputProps
  extends UseControllerProps,
    Omit<PickerProps, 'ref' | 'date'> {
  label?: string;
  labelStyle?: StyleProp<TextStyle>;
  errorStyle?: StyleProp<TextStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  rightChild?: JSX.Element;
}

export function AppDateInput({
  control,
  name,
  rules,
  label,
  labelStyle,
  errorStyle,
  containerStyle,
  defaultValue = '',
  mode = 'date',
  modal = true,
  minimumDate = new Date('1900-01-01'),
  ...pickerProps
}: AppDateInputProps) {
  const [openModal, setOpenModal] = React.useState(false);

  const {errors} = useFormState({control, name});

  return (
    <View style={containerStyle}>
      {label && <AppText style={labelStyle}>{label}</AppText>}

      <Controller
        defaultValue={defaultValue}
        name={name}
        control={control}
        rules={rules}
        render={({field: {onChange, value}}) => (
          <View>
            <Pressable
              onPress={() => {
                setOpenModal(true);
              }}
              style={[styles.textContainer, {borderColor: Colors.border}]}>
              {value ? (
                <Text style={[styles.text, {color: Colors.text}]}>
                  {dayjs(value).format('YYYY-MM-DD')}
                </Text>
              ) : (
                <Text style={[styles.text, {color: Colors.placeholder}]}>
                  YYYY-MM-DD
                </Text>
              )}
            </Pressable>

            <DatePicker
              open={openModal}
              theme="light"
              modal={modal}
              minimumDate={minimumDate}
              mode={mode}
              date={value || new Date()}
              title={null}
              onConfirm={date => {
                setOpenModal(false);
                onChange(date);
              }}
              onCancel={() => {
                setOpenModal(false);
              }}
              {...pickerProps}
            />
          </View>
        )}
      />

      {errors[name] && !!errors[name]?.message && (
        <AppText
          style={[
            {
              color: Colors.error,
            },
            errorStyle,
          ]}>
          {errors[name]?.message as string}
        </AppText>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  text: {
    flex: 1,
    fontSize: Sizes.regular,
    paddingVertical: Sizes.padding,
    paddingHorizontal: Sizes.paddinglx,
    justifyContent: 'space-between',
  },
  textContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    borderWidth: Sizes.borderWidth,
    borderRadius: Sizes.borderRadius,
    justifyContent: 'space-between',
    overflow: 'hidden',
  },
});
