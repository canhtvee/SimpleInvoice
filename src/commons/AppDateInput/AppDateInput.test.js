import React from 'react';
import {useForm} from 'react-hook-form';
import {cleanup, renderHook} from '@testing-library/react-native';
import {create} from 'react-test-renderer';

import {AppDateInput} from './AppDateInput';

/**
 * Run test command
 * yarn test -u -t="AppDateInput"
 */

afterEach(cleanup);

describe('AppDateInput', () => {
  test('should render correctly', () => {
    const {result} = renderHook(() => {
      return useForm();
    });

    const {control} = result.current;

    const tree = create(
      <AppDateInput
        control={control}
        name={'test-input'}
        label={'test-label'}
        placeholder={'test-input'}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();

    expect(1).toBeDefined();
  });
});
