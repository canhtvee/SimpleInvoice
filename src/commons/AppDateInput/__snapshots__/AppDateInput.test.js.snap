// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppDateInput should render correctly 1`] = `
<View>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 27.5,
        },
        undefined,
      ]
    }
  >
    test-label
  </Text>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 6,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "overflow": "hidden",
          },
          {
            "borderColor": "#0000008A",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "flex": 1,
              "fontSize": 27.5,
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 23.5,
            },
            {
              "color": "#00000061",
            },
          ]
        }
      >
        YYYY-MM-DD
      </Text>
    </View>
  </View>
</View>
`;
